### ISSUE TITLE
Using a variable with the type: list[ <dataclass> | None ], Pylance reports problem with valid syntax in if statement

### ISSUE THREAD

**Issue by sstreet98** [microsoft/pyright#10665]
Pylance version: 2025.6.2
 OS and version: Debian Linux 12
 Python version: 3.11.2

## Code Snippet

```python
from dataclasses import dataclass
import random


@dataclass
class Data:
    foo:int
    bar:str


varlist : list[Data|None] = []

for index in range(0, 10):
    varlist.append(None)

ranindex = random.randint(0,9)
print(f"Setting {ranindex} to BBSClient")
varlist[ranindex] = Data(foo=1,bar="broken")


for index in range(0, 10):
    print(f"Checking index {index}...")
    if varlist[index] and varlist[index].foo == 1 :  # pyLance reports: "foo" is not a known attribute of "None"
        print(f"{index} has foo=1")
    else:
        print(f"does not have foo=1")

    if varlist[index] is not None and varlist[index].foo == 1: # pyLance reports: "foo" is not a known attribute of "None"
        pass

    if type(varlist[index]) is Data and varlist[index].foo == 1:  # pyLance reports: "foo" is not a known attribute of "None"
        pass
```

## Repro Steps

Load this code snippet into VSCode with Pylance enabled,  see that "foo" is underlined in red in each of the if statements.   However, this code runs and works as expected.

## Expected behavior

foo should be a valid attribute of the "Data" class

## Actual behavior

foo is marked as an unknown attribute of the "None" class

## Logs
<none needed, very easy to reproduce>

**Comment by erictraut** [comment]
Pyright (the type checker that pylance is built upon) is working as designed here, so I don't consider this a bug.

Pyright is not reporting a problem with the syntax of your code. It's reporting a type violation. Pyright checks for type violations by statically analyzing your code to determine the types of specific expressions. For details, you can read more about [the basics of static typing](https://microsoft.github.io/pyright/#/type-concepts) in the pyright documentation.

You'll also want to make sure you understand the concept of "type narrowing", which is explained [here](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-narrowing). Type narrowing allows a static type checker to "narrow" (or "refine") the type of certain expressions based on conditional tests ("type guards") in your code. Narrowing can be applied safely only to certain expression forms. In your example, you're using a subscript expression with a variable (dynamic) subscript, and that cannot be safely narrowed.

If you want your code to type check without errors, here's the way I recommend you change it:

```python
    if (val := varlist[index]) and val.foo == 1: ...
```

By assigning the subscripted value to a temporary variable, you not only guarantee type safety, but it's also faster at runtime because you're not calling `__getitem__` redundantly.

**Comment by sstreet98** [comment]
Thank you for the recommendation and resolution.
